libopeniscsiusr/iface.c

```diff
@@ -268,6 +267,52 @@ int _iscsi_iface_get_from_sysfs(struct iscsi_context *ctx, uint32_t host_id,
 	return rc;
 }
 
+/* create all ifaces for a host from sysfs */
+int _iscsi_ifaces_get_from_sysfs(struct iscsi_context *ctx, uint32_t host_id,
+				 struct iscsi_iface ***ifaces, uint32_t *iface_count)
+{
+	int rc = LIBISCSI_OK;
+	char **iface_kern_ids = NULL;
+
+	assert(ctx != NULL);
+	assert(ifaces != NULL);
+
+	*ifaces = NULL;
+	*iface_count = 0;
+
+	_good(_iscsi_iface_kern_ids_of_host_id(ctx, host_id, &iface_kern_ids, iface_count),
+	      rc, out);
+	if (*iface_count > 0) {
+		*ifaces = (struct iscsi_iface **) calloc(*iface_count,
+							 sizeof(struct iscsi_iface *));
+		_alloc_null_check(ctx, *ifaces, rc, out);
+		for (uint32_t i = 0; i < *iface_count; i++) {
```

Comment:
Got compile failure on RHEL 7.5:
```
sysfs.c: In function ‘_iscsi_iface_kern_ids_of_host_id’:
sysfs.c:502:2: error: ‘for’ loop initial declarations are only allowed in C99 mode
  for (uint32_t i = 0; i < *iface_count; i++) {
  ^
sysfs.c:502:2: note: use option -std=c99 or -std=gnu99 to compile your code
sysfs.c:510:3: error: ‘for’ loop initial declarations are only allowed in C99 mode
   for (uint32_t i = 0; i < *iface_count; i++ ) {
   ^
cc1: all warnings being treated as errors
make: *** [sysfs.o] Error 1
```

Reply:
This brings up a larger conversation about language standards, compiler extensions, and coding style.  We should probably make some choices about what to allow, and then get the CFLAGS set to match.  I'm not too hung up on loop initial declarations, but in general I think allowing at least C99 can result in much cleaner code.

Related but separate, I was working up a patch to use asprintf instead of snprintf to the stack and removing all PATH_MAX buffers.  asprintf is non-standard, but we only care about Linux and it's supported in glibc and musl.

In the process, I also started using the gcc __attribute__((__cleanup__(f))) to make it obvious which allocations are local to the function and get the compiler to emit code to always free them as they leave scope.  I may have gotten use to the idea when working on systemd code, and think it makes C much nicer to deal with (with some helper cleanup functions and macros to make the attribute syntax easier to look at).